enclave-runtime/src/state.rs

```diff
@@ -15,68 +15,110 @@
 
 */
 
-use crate::{error::Result, hex, io, utils::UnwrapOrSgxErrorUnexpected};
+use crate::{
+	error::{Error, Result},
+	hex, io,
+	utils::UnwrapOrSgxErrorUnexpected,
+};
 use base58::{FromBase58, ToBase58};
 use codec::{Decode, Encode};
 use ita_stf::{ShardIdentifier, State as StfState, StateType as StfStateType, Stf};
 use itp_settings::files::{ENCRYPTED_STATE_FILE, SHARDS_PATH};
 use itp_sgx_crypto::{AesSeal, StateCrypto};
 use itp_sgx_io::SealedIO;
+use lazy_static::lazy_static;
 use log::*;
 use sgx_tcrypto::rsgx_sha256_slice;
 use sgx_types::*;
 use sp_core::H256;
-use std::{fs, io::Write, path::Path, vec::Vec};
+use std::{
+	fs,
+	io::Write,
+	path::Path,
+	sync::{SgxRwLock as RwLock, SgxRwLockWriteGuard as RwLockWriteGuard},
+	vec::Vec,
+};
 
 /// Facade for handling STF state from file
 pub trait HandleState {
-	/// Load the STF state for a specific shard
+	type WriteLockPayload;
+
+	/// Load the state for a given shard
+	///
+	/// Initializes the shard and state if necessary, so this is guaranteed to
+	/// return a state
 	fn load_initialized(&self, shard: &ShardIdentifier) -> Result<StfState>;
 
+	fn load_for_mutation(
+		&self,
+		shard: &ShardIdentifier,
+	) -> Result<(RwLockWriteGuard<'_, Self::WriteLockPayload>, StfState)>;
+
 	/// Writes the state (without the state diff) encrypted into the enclave
 	///
 	/// Returns the hash of the saved state (independent of the diff!)
-	fn write(&mut self, state: StfState, shard: ShardIdentifier) -> Result<H256>;
+	fn write(
+		&self,
+		state: StfState,
+		state_lock: RwLockWriteGuard<'_, Self::WriteLockPayload>,
+		shard: &ShardIdentifier,
+	) -> Result<H256>;
 
 	/// Query whether a given shard exists
 	fn exists(&self, shard: &ShardIdentifier) -> bool;
 
-	/// Initialize a shard with a given identifier
-	fn init_shard(&mut self, shard: &ShardIdentifier) -> Result<()>;
-
 	/// List all available shards
 	fn list_shards(&self) -> Result<Vec<ShardIdentifier>>;
 }
 
-pub struct StateFacade;
+lazy_static! {
+	// as long as we have a file backend, we use this 'dummy' lock,
+	// which guards against concurrent read/write access
+	pub static ref STF_STATE_LOCK: RwLock<()> = Default::default();
+}
```

Comment:
The global lock we use for accessing the state. Once we move to an in-memory state, we will have a 'proper' lock that wraps the state itself.